File: integrations/src/main/java/com/nivasafinance/integrations/framework/runner/ServiceRunner.java
package com.nivasafinance.integrations.framework.runner;

import com.nivasafinance.integrations.framework.config.BusinessContext;
import com.nivasafinance.integrations.framework.core.data.RunConfig;
import com.nivasafinance.integrations.framework.core.exception.NavigatorIntegrationClientException;
import com.nivasafinance.integrations.framework.core.exception.NavigatorIntegrationServerException;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public class ServiceRunner<S, A> {

    private final S primary;
    private final S fallback;
    private final int retries;

    public ServiceRunner(S primary, S fallback, int retries) {
        this.primary = primary;
        this.fallback = fallback;
        this.retries = retries;
    }

    public Object invokeService(
            String methodName,
            A argument,
            RunConfig runConfig,
            BusinessContext businessContext) throws NavigatorIntegrationClientException, NavigatorIntegrationServerException {
        NavigatorIntegrationServerException lastServerException = null;

        for (int i = 0; i < retries; i++) {
            try {
                Object result = tryPrimaryService(methodName, argument, runConfig, businessContext);
                if (result != null) {
                    return result;
                }
            } catch (NavigatorIntegrationServerException e) {
                lastServerException = e;
            }
        }

        return tryFallbackService(methodName, argument, runConfig, businessContext, lastServerException);
    }

    @SuppressWarnings("unchecked")
    private Object tryPrimaryService(
            String methodName,
            A argument,
            RunConfig runConfig,
            BusinessContext businessContext) throws NavigatorIntegrationServerException, NavigatorIntegrationClientException {
        try {
            Class<?> argumentClass = argument != null ? argument.getClass() : null;
            Method method = primary.getClass().getMethod(
                    methodName,
                    argumentClass,
                    runConfig.getPrimaryConfig().getClass(),
                    BusinessContext.class
            );
            return method.invoke(primary, argument, runConfig.getPrimaryConfig(), businessContext);
        } catch (InvocationTargetException ex) {
            Throwable cause = ex.getTargetException();
            if (cause instanceof NavigatorIntegrationServerException) {
                throw new NavigatorIntegrationServerException(cause.getMessage());
            } else {
                throw new NavigatorIntegrationClientException(cause.getMessage());
            }
        } catch (NavigatorIntegrationServerException ex) {
            throw new NavigatorIntegrationServerException("Server error: " + ex.getMessage());
        } catch (NavigatorIntegrationClientException ex) {
            throw new NavigatorIntegrationClientException("Client error: " + ex.getMessage());
        } catch (Exception ex) {
            throw new NavigatorIntegrationClientException("Error invoking service: " + ex.getMessage());
        }
    }

    @SuppressWarnings("unchecked")
    private Object tryFallbackService(
            String methodName,
            A argument,
            RunConfig runConfig,
            BusinessContext businessContext,
            NavigatorIntegrationServerException lastServerException) throws NavigatorIntegrationServerException, NavigatorIntegrationClientException {
        try {
            if (fallback == null || runConfig.getFallbackConfig() == null) {
                if (lastServerException != null) {
                    throw lastServerException;
                }
                return null;
            }
            Class<?> argumentClass = argument != null ? argument.getClass() : null;
            Method method = fallback.getClass().getMethod(
                    methodName,
                    argumentClass,
                    runConfig.getFallbackConfig().getClass(),
                    BusinessContext.class
            );
            return method.invoke(fallback, argument, runConfig.getFallbackConfig(), businessContext);
        } catch (NavigatorIntegrationServerException ex) {
            throw new NavigatorIntegrationServerException("Server error: " + ex.getMessage());
        } catch (NavigatorIntegrationClientException ex) {
            throw new NavigatorIntegrationClientException("Client error: " + ex.getMessage());
        } catch (Exception ex) {
            if (lastServerException != null) {
                throw lastServerException;
            }
            return null;
        }
    }
}

File: integrations/src/main/kotlin/framework/runner/ServiceRunner.kt
package framework.runner

import framework.config.BusinessContext
import framework.core.data.RunConfig
import framework.core.exception.NavigatorIntegrationClientException
import framework.core.exception.NavigatorIntegrationServerException
import java.lang.reflect.InvocationTargetException

class ServiceRunner<S, A>(
    val primary: S,
    val fallback: S,
    var retries: Int,
) {
    @Throws(NavigatorIntegrationClientException::class, NavigatorIntegrationServerException::class)
    fun invokeService(
        methodName: String,
        argument: A?,
        runConfig: RunConfig,
        businessContext: BusinessContext
    ): Any? {
        val lastServerException: NavigatorIntegrationServerException? = null

        repeat(retries) {
            val result = tryPrimaryService(methodName, argument, runConfig, businessContext)
            if (result != null) return result
        }

        return tryFallbackService(methodName, argument, runConfig, businessContext, lastServerException)
    }

    @Suppress("ThrowsCount", "SwallowedException", "UnsafeCallOnNullableType")
    private fun tryPrimaryService(
        methodName: String,
        argument: A?,
        runConfig: RunConfig,
        businessContext: BusinessContext
    ): Any? {
        return try {
            val method = primary!!::class.java.getMethod(
                methodName,
                argument?.javaClass,
                runConfig.primaryConfig::class.java,
                BusinessContext::class.java
            )
            method.invoke(primary, argument, runConfig.primaryConfig, businessContext)
        } catch (ex: InvocationTargetException) {
            val cause = ex.targetException
            if (cause is NavigatorIntegrationServerException) {
                throw NavigatorIntegrationServerException(cause.message)
            } else {
                throw NavigatorIntegrationClientException(cause.message)
            }
        } catch (ex: NavigatorIntegrationServerException) {
            throw NavigatorIntegrationServerException("Server error: ${ex.message}")
        } catch (ex: NavigatorIntegrationClientException) {
            throw NavigatorIntegrationClientException("Client error: ${ex.message}")
        }
    }

    @Suppress("ThrowsCount", "SwallowedException", "UnsafeCallOnNullableType")
    private fun tryFallbackService(
        methodName: String,
        argument: A?,
        runConfig: RunConfig,
        businessContext: BusinessContext,
        lastServerException: NavigatorIntegrationServerException?
    ): Any? {
        return try {
            val method = fallback!!::class.java.getMethod(
                methodName,
                argument?.javaClass,
                runConfig.fallbackConfig?.javaClass,
                BusinessContext::class.java
            )
            method.invoke(fallback, argument, runConfig.fallbackConfig, businessContext)
        } catch (ex: NavigatorIntegrationServerException) {
            throw NavigatorIntegrationServerException("Server error: ${ex.message}")
        } catch (ex: NavigatorIntegrationClientException) {
            throw NavigatorIntegrationClientException("Client error: ${ex.message}")
        } catch (_: Exception) {
            lastServerException?.let { throw it }
            null
        }
    }
}
